[RFC] Fix for type deduction logic for forwarding references

shwina opened the pull request:
The logic previously added for type deduction involving forwarding references breaks down in some situations when we have other overloads. This PR fixes that.

shwina commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -5784,18 +5784,8 @@ def analyse_c_function_call(self, env):
-                            arg_types[i] = PyrexTypes.c_ref_type(arg_types[i])
-
-            entry = PyrexTypes.best_match(arg_types, alternatives, self.pos, env, args)
+            entry = PyrexTypes.best_match([arg.type for arg in args],
```
I moved this logic to inside `best_match`, so that we don't "permanently" alter the argument types.

shwina commented on Cython/Compiler/PyrexTypes.py:
```diff
@@ -4599,12 +4599,22 @@ def best_match(arg_types, functions, pos=None, env=None, args=None):
+            # function call argument is an lvalue. See:
+            # https://en.cppreference.com/w/cpp/language/template_argument_deduction#Deduction_from_a_function_call
+            arg_types_for_deduction = arg_types.copy()
+            if func.type.is_cfunction and args:
```
`args` could be `None` here, so I'm not 100% sure about this fix...

shwina commented:
@scoder There are several places where [`best_match`](https://github.com/cython/cython/blob/master/Cython/Compiler/PyrexTypes.py#L4549) is called without the `args` parameter. There's even a [TODO](https://github.com/cython/cython/blob/master/Cython/Compiler/PyrexTypes.py#L4571) to remove this parameter entirely and replace it with a list of types (currently, it's a list of nodes).

This is a problem for things like forwarding references because deduction depends not only on the type of the argument but also its _expression type_ (i.e., whether it's an lvalue expression or rvalue expression):

```cython
cdef foo[T](T&& x)

cdef int x = 1
cdef foo(x)  # should generate foo<int&>(x);
cdef foo(1)  # should generate foo<int>(1);
```

The tests in this PR are passing but it feels a bit fragile to me. Is there a more elegant solution I may be missing? :-)

scoder commented on Cython/Compiler/PyrexTypes.py:
```diff
@@ -4599,12 +4599,24 @@ def best_match(arg_types, functions, pos=None, env=None, args=None):
+            # use lvalue-reference-to-A for deduction in place of A when the
+            # function call argument is an lvalue. See:
+            # https://en.cppreference.com/w/cpp/language/template_argument_deduction#Deduction_from_a_function_call
+            arg_types_for_deduction = copy.copy(arg_types)
```
Let's keep this simple.
```suggestion
            arg_types_for_deduction = list(arg_types)
```

scoder reviewed (commented):
If we now _require_ `args` to be a list of nodes, then we should state that clearly and also change the TODO comment.
I think this is a net-improvement, even if there might be cases where it still fails to to a missing `args`. If those still exist, they will eventually have to fixed in one way or another.

shwina commented:
I removed the `TODO` and added a note that the `args` parameter is expected to be a list of `ExprNodes`.

shwina commented:
@scoder could you take another look? Thank you!

shwina commented:
Could I request another review of this PR? cc: @da-woods

da-woods commented:
Hi @shwina - I hadn't commented on this because it looked pretty good to me.

If I was forced to make a suggestion, I'd say it might be worth testing a few more corner cases. Things like does it fail elegantly if you don't provide enough args to deduce the type (or args is None):

```
cdef extern from *:
    cdef one_arg[T](T&&)

one_arg()  # clearly not going to work, but hopefully doesn't break Cython
```

does overriding the types manually work? (e.g. `one_arg[int](1)`, `cdef int x=1; one_arg[int&](x)`, and `one_arg[int&&](1)` look like they should work by analogy with C++).

shwina commented:
Thanks @da-woods, and apologies - I didn't mean to put you on the spot or anything. Just wanted to get another review after addressing the concerns above and get the ball rolling on this PR again :-)

Will add more tests as suggested.

shwina commented:
> Things like does it fail elegantly if you don't provide enough args to deduce the type (or args is None):

I added tests for too few and too many arguments and they work as expected. However:

> does overriding the types manually work? 

At the moment, no -- because the parser code currently doesn't know to parse the `int&` or `int&&` in the expressions below:

```
one_arg[int&](x)
one_arg[int&&](x)
```

I'd agree that these should work by analogy with C++ -- do you think I should tackle that in this PR, or would you prefer that I open another one? At first glance, it seems like a non-trivial amount of work that would be somewhat orthogonal to the intent of this PR.

da-woods commented:
> At first glance, it seems like a non-trivial amount of work that would be somewhat orthogonal to the intent of this PR.

I think that's probably fair enough and it doesn't need to be in this PR

shwina commented:
@scoder could this PR please be considered for merging? Thanks :)

shwina commented:
Closing in favor of https://github.com/cython/cython/pull/3790 where I will just include these changes.
